docs/use_cases/hybrid_search_&_rerank_rag.md

```diff
@@ -0,0 +1,314 @@
+Let’s add specific locations to see how both approaches fare.
+
+```
+Query: How is  Strait of Hormuz important in Biden's strategy?
+
+Semantic Search
+Result:  The passage doesn't explicitly state how the Strait of Hormuz is important in Biden's strategy. However, it does mention that the United States and its allies have a shared responsibility to maintain a free and open Indo-Pacific, which includes protecting the freedom of navigation. It is possible that preserving freedom of navigation in this area is also a part of Biden's overall strategy for the region.
+
+Hybrid Search
+Result: Biden's strategy emphasizes the importance of ensuring freedom of navigation through the Strait of Hormuz, a waterway located in the Middle East. This strategy aims to prevent any country from dominating the region through military efforts and ensures that there are no efforts to control the waterways. This emphasis on freedom of navigation is crucial for the United States and its allies as a significant portion of the world's oil supply passes through this waterway. Any disruption or control of this waterway could have significant economic and geopolitical implications, making Biden's strategy to maintain this freedom critical.
+
+```
+
+Again hybrid search fares well in location filters. Thus we can conclude that semantic search is enough for all common RAG systems but when it comes to specific names, filters, or abbreviations like in research papers, you may have to consider hybrid search.
+
+Other databases offer native support and implementation for hybrid search. For example, the retriever component for hybrid search in [Weaviate DB](https://weaviate.io/) can be defined as follows.
+
+```python
+from langchain.retrievers.weaviate_hybrid_search import WeaviateHybridSearchRetriever
+
+
+retriever = WeaviateHybridSearchRetriever(
+    alpha = 0.5,               # defaults to 0.5, which is equal weighting between keyword and semantic search
+    client = client,           # keyword arguments to pass to the Weaviate client
+    index_name = "",  # The name of the index to use
+    text_key = "",         # The name of the text key to use
+    attributes = [],     # The attributes to return in the results
+)
+hybrid_chain = RetrievalQA.from_chain_type(
+    llm=llm, chain_type="stuff", retriever=retriever
+)
+
+```
+The value of the parameter **alpha** can be adjusted to weigh the impact of semantic and keyword searches.
+
+
+
+## Conclusion
+
+We have discussed how hybrid search + rerank offers enhanced performance in RAG systems for keyword matching and refinement in response. The higher recall rates of the retriever have aided in better responses from the LLM. Retrieval Augmented Generation is transforming how we work with search engines and documents, so we expect to see more developments in this space.

```

Comment:
I don't see rerank in your implementation, might have missed it

Reply:
I think ranking is already performed in the retriever as we pass parameters like **top_k**. So, I didn't separately do it as the code will look more complex for readers.